fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/expressions/Comparisons.java

```diff
@@ -242,8 +242,61 @@ private static Boolean compareEquals(Object value, Object comparand) {
         if (value == null || comparand == null) {
             return null;
         } else {
-            return toClassWithRealEquals(value).equals(toClassWithRealEquals(comparand));
+            if (value instanceof Message) {
+                if (!(comparand instanceof Message)) {
+                    return false;
+                }
+                return compareMessageEquals((Message)value, (Message)comparand);
+            } else {
+                return toClassWithRealEquals(value).equals(toClassWithRealEquals(comparand));
+            }
+        }
+    }
+
+    // return true if m1 and m2 are equal, i.e. if for all field index i, m1.getFieldByIndex(i) = m2.getFieldByIndex(i), m1 and m2 are equal
+    // m1 and m2 must only have primitive types (not repeated, map, or message type fields)
+    private static boolean compareMessageEquals(@Nonnull Message m1, @Nonnull Message m2) {
```

Comment:
This code should bo to `MessageHelpers`. I am actually surprised that such a  method does not exist yet. It should also operate on ordinal positioning inside of the proto message as field indexes do not matter here.